### PR TITLE
feature: Dynamic plugin loader

### PR DESCRIPTION
**Pull request by davy-tw**
We changed the concept of "plugin", for now, plugin is a npm package which exports in commonjs2[^1] format.

And plugins system will scan plugins placed in `/usr/share/einstein/plugins` (non-windows only) and `$HOME/.config/einstein/plugins`.

Plugins must follow these rules:

## Declare `name`, `uid` and the entrypoint (`main`/`module`) in `package.json`:
```json
{
  "name": "example",
  "version": "0.0.1",
  "uid": "tw.childish.einstein.plugins.example",
  "module": "index.js" // or `"main": "index.js"`
}
```

Since `uid` is unique identifier of the plugin, using `<domain>.<name>` is recommended

## Use `fuse.js`[^2] as `peerDependency` if needed

Since fuzzy-searching is highly usage in a searchable application,
we are going to share `fuse.js` with plugins, so you don't need to bring this by your own.

If you're using webpack, set `externals['fuse.js']` to `'fuse.js'`.

## Export in CommonJS2 format

For now, we only support to load plugins which is in commonjs2 format.

If you're using webpack, set `output.libraryTarget` to `'commonjs2'`.

# Progress

- [x] Implement PluginScanner to scan plugins
- [ ] Implement PluginLoader to load plugins more securely
- [ ] Write documents

--

[^1]: Which used in Node.js commonly, see: https://github.com/webpack/webpack/issues/1114
[^2]: A fuzzy searcher widely used in einstein project, see: https://fusejs.io/

### CODE REVIEW

File: plugins/pass/index.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { PluginContext, PluginSetup } from 'einstein'
+
+import PassSearchEngine from './PassSearchEngine'
+
+const setup: PluginSetup = async (context: PluginContext) => {
+	const searchEngine = new PassSearchEngine()
+
+	context.registerSearchEngine(searchEngine, ...PassSearchEngine.triggers)
+	context.registerEventHandler('pass', (data: any) => searchEngine.copyPassword(data))
```
Comment (xatier) [inline plugins/pass/index.ts:9]: suggestion: add

```typescript
type EventType = 'pass' | 'pass show'
```

File: src/api/types.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export type UID = string
```
Comment (xatier) [inline src/api/types.ts:2]: suggestion: add `type EventType = string` here

File: src/main/index.ts [hunk 3]
```diff
@@ -21,17 +21,21 @@ const registerMessageTunnelPair = <T = any>(
 }
 
 const createApp = async () => {
-	const { exitProcess: closePluginHost, messageTunnel: pluginHostTunnel } = await usePluginHost()
 	const { window: sharedProcessWindow } = await useSharedProcess()
+	console.log('SharedProcess is ready.')
+
+	const { exitProcess: closePluginHost, messageTunnel: pluginHostTunnel } = await usePluginHost()
 
 	const pluginIsReady = new Promise<void>((resolve) => {
```
Comment (xatier) [inline src/main/index.ts]: issue: remove `pluginIsReady`, inline `await`.

File: src/main/index.ts [hunk 4]
```diff
@@ -21,17 +21,21 @@ const registerMessageTunnelPair = <T = any>(
 }
 
 const createApp = async () => {
-	const { exitProcess: closePluginHost, messageTunnel: pluginHostTunnel } = await usePluginHost()
 	const { window: sharedProcessWindow } = await useSharedProcess()
+	console.log('SharedProcess is ready.')
```
Comment (xatier) [inline src/main/index.ts:25]: suggestion: `has been loaded`

File: src/main/useOmniSearch.ts [hunk 5]
```diff
@@ -26,6 +26,7 @@ const createWindow = () => {
 		show: false,
 		alwaysOnTop: true,
 		webPreferences: {
+			contextIsolation: false,
```
Comment (xatier) [inline src/main/useOmniSearch.ts]: suggestion: add a `TODO` comment here

File: src/main/useSharedProcess.ts [hunk 6]
```diff
@@ -18,6 +18,7 @@ const createSharedProcess = () => {
 	const window = new BrowserWindow({
 		show: false,
 		webPreferences: {
+			contextIsolation: false,
```
Comment (xatier) [inline src/main/useSharedProcess.ts]: suggestion: add a `TODO` comment here

File: src/pluginHost.node/PluginManager.ts [hunk 7]
```diff
@@ -1,52 +1,146 @@
...
+		} catch (e) {
+			console.log(e.message)
 		}
 	}
 
-	register(plugin: IPlugin): this {
-		if (this.plugins.find(({ uid }) => uid === plugin.uid)) {
-			return this // duplicated
+	async unloadPlugin(uid: UID) {
+		const plugin = this.plugins.get(uid)
```
Comment (xatier) [inline src/pluginHost.node/PluginManager.ts:80]: issue: add `undefined` guard.

File: src/pluginHost.node/PluginManager.ts [hunk 8]
```diff
@@ -1,52 +1,146 @@
...
+		const eventHandlers: Record<string, Set<PluginEventHandler>> = {}
+
+		return {
+			get eventHandlers() {
+				return eventHandlers
+			},
+
+			registerEventHandler: (type: string, handler: PluginEventHandler) => {
+				if (!eventHandlers[type]) {
+					eventHandlers[type] = new Set()
```
Comment (xatier) [inline src/pluginHost.node/PluginManager.ts:102]: suggestion: add generic type here

File: src/pluginHost.node/PluginManager.ts [hunk 9]
```diff
@@ -1,52 +1,146 @@
...
+}
+const hasPluginUID = <T>(obj: T): obj is WithPluginUID<T> => {
+	return Object.prototype.hasOwnProperty.call(obj, PluginUIDSymbol)
+}
 
 class PluginManager {
-	private plugins: IPlugin[]
+	private plugins: Map<UID, Plugin> = new Map()
+
+	private triggerMap: Map<string, Set<ISearchEngine>> = new Map([ [ VOID_TRIGGER, new Set() ] ])
```
Comment (xatier) [inline src/pluginHost.node/PluginManager.ts:53]: suggestion: use `string | Symbol` here

File: src/pluginHost.node/PluginManager.ts [hunk 10]
```diff
@@ -56,43 +150,45 @@ class PluginManager {
...
+		const plugin = this.plugins.get(uid)
 		if (!plugin) {
-			return []
+			return
 		}
 
-		const engine = plugin.searchEngines.find(({ name }) => name === engineName)
-		if (!engine) {
-			return []
+		const handlers = plugin!.eventHandlers[type]
```
Comment (xatier) [inline src/pluginHost.node/PluginManager.ts]: suggestion: consider removing assertion `!`.

File: src/pluginHost.node/PluginScanner.ts [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+					.map((path) => this.readPluginMetadata(path)),
+			)
+		} catch (e) {
+			console.log(`Unable to readdir: ${dir}, reason: ${e.message}`)
+			return []
+		}
+	}
+
+	private async readPluginMetadata(path: string): Promise<PluginMetadata | null> {
+		console.log(`Scanned plugin path: ${path}`)
```
Comment (xatier) [inline src/pluginHost.node/PluginScanner.ts]: suggestion: remove this log.

File: src/pluginHost.node/PluginMetadata.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import { UID } from 'einstein'
+
+export type PluginMetadata = {
+	name: string
+	uid: UID
+	entry: string
+	path?: string
```
Comment (xatier) [inline src/pluginHost.node/PluginMetadata.ts]: suggestion: remove optional `?`

**Review by xatier — COMMENTED** [review]
gogo

File: src/api/types.ts (hunk 2 above)
Comment (davy-tw) [reply src/api/types.ts:2]: done in c1476cd

Comment (xatier) [reply src/api/types.ts:2]: lgtm.

File: plugins/pass/index.ts (hunk 1 above)
Comment (davy-tw) [reply plugins/pass/index.ts:9]: I think improving plugins is not the objective of this PR, maybe we can add this in next step?

File: src/main/useOmniSearch.ts (hunk 5 above)
Comment (davy-tw) [reply src/main/useOmniSearch.ts]: added in 108c06e6fb36ca95212aeef26ead44f0556b3aad

File: src/main/useSharedProcess.ts (hunk 6 above)
Comment (davy-tw) [reply src/main/useSharedProcess.ts]: added in 108c06e6fb36ca95212aeef26ead44f0556b3aad

File: plugins/example/index.ts [hunk 13]
```diff
@@ -25,8 +23,7 @@ const searchEngine: ISearchEngine = {
 	},
 }
 
-const searchEngine2: ISearchEngine = {
-	name: 'voidEngine',
+const searchEngine2 = {
 	triggers: [ VOID_TRIGGER ],
```
Comment (xatier) [inline plugins/example/index.ts]: suggestion: remove this

Comment (davy-tw) [inline plugins/example/index.ts]: remove this line, since `VOID_TRIGGER` is default trigger

**Pull request merged**